clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -294,18 +365,33 @@ private void completeBatch(RecordBatch batch, ProduceResponse.PartitionResponse
         if (error != Errors.NONE && canRetry(batch, error)) {
             // retry
             log.warn("Got error produce response with correlation id {} on topic-partition {}, retrying ({} attempts left). Error: {}",
-                     correlationId,
-                     batch.topicPartition,
-                     this.retries - batch.attempts - 1,
-                     error);
-            this.accumulator.reenqueue(batch, now);
-            this.sensors.recordRetries(batch.topicPartition.topic(), batch.recordCount);
+                    correlationId,
+                    batch.topicPartition,
+                    this.retries - batch.attempts - 1,
+                    error);
+            if (transactionState == null || (transactionState.pid() == batch.producerId())) {
+                // If idempotence is enabled only retry the request if the current PID is the same as the pid of the
+                // batch.
+                this.accumulator.reenqueue(batch, now);
+                this.sensors.recordRetries(batch.topicPartition.topic(), batch.recordCount);
+            }
         } else {
             RuntimeException exception;
             if (error == Errors.TOPIC_AUTHORIZATION_FAILED)
                 exception = new TopicAuthorizationException(batch.topicPartition.topic());
             else
                 exception = error.exception();
+
+            if (error == Errors.OUT_OF_ORDER_SEQUENCE_NUMBER && batch.producerId() == transactionState.pid()) {
+                // Reset the transactional state and get a new pid, since we can't recover from an out of order
+                // sequence error.
+                transactionState.reset();
+                while (!transactionState.hasPid()) {
```

Comment:
This is the same snippet we have above. Maybe it should be a method? I'm wondering if we should just add a flag indicating that a PID is needed and then add a check for the flag in the `run` loop? Then we would avoid blocking here before invoking the callback.